cpp/src/arrow/dataset/file_test.cc

```diff
@@ -349,18 +349,17 @@ TEST_F(TestFileSystemDataset, WriteProjected) {
 }
 
 class FileSystemWriteTest : public testing::TestWithParam<std::tuple<bool, bool>> {
+  using PlanFactory = std::function<cp::Declaration(
+      const FileSystemDatasetWriteOptions&,
+      std::function<Future<util::optional<cp::ExecBatch>>()>*)>;
+
```

Comment:
I believe the concatenation code in this particular test could be even simpler if `PlanFactory` had a return-type of `std::vector<Declaration>`, because `Sequence` would do the checking-and-wiring work when passed the concatenated vector. OTOH, the `Root` and `Concat` functions you added to `Declaration` seem generally useful, so this test code is good for covering it. I'll let you choose the alternative you prefer.

Reply:
Given the ["No clear root" test error](https://github.com/apache/arrow/runs/6395740926?check_suite_focus=true), maybe the easiest fix would be the former alternative.

Reply:
I did think about this but I would like to start generally improving our test code for exec plans.  I think we are a little sparse at the moment and it is rather complex to create some of these unit tests.  I'll try and wrap this up tomorrow and if not I can leave the bonus utilities for a future PR.